py4cast/datasets/poesy.py

```diff
@@ -256,45 +330,25 @@ def sample_list(self):
         Create a list of sample from information
         """
         print("Start forming samples")
-        terms = list(
-            np.arange(
-                self.settings.term["start"],
-                self.settings.term["end"],
-                self.settings.term["timestep"],
-            )
-        )
-
-        sample_by_date = len(terms) // self.settings.num_total_steps
 
         samples = []
-        number = 0
+        n_samples = 0
 
-        for date in self.period.date_list:
-            for member in self.settings.members:
-                for sample in range(0, sample_by_date):
-                    input_terms = terms[
-                        sample
-                        * self.settings.num_total_steps : sample
-                        * self.settings.num_total_steps
-                        + self.settings.num_input_steps
-                    ]
-                    output_terms = terms[
-                        sample * self.settings.num_total_steps
-                        + self.settings.num_input_steps : sample
-                        * self.settings.num_total_steps
-                        + self.settings.num_input_steps
-                        + self.settings.num_output_steps
-                    ]
-                    samp = Sample(
-                        date=date,
-                        member=member,
-                        input_terms=input_terms,
-                        output_terms=output_terms,
-                    )
+        # Run through the dataset and get a list of sample's args.
+        list_args_sample = run_through_timestamps(self.period, self.settings)
```

Comment:
**run_through_timestamps** returns a **List[Dict[str, Any]]**, this is an implicit and blur interface. On the other hand your Poesy samples expect real known and explicit args with a known type (**date, member, input_terms, output_terms**)

Maybee you need an intermediate explicit **SampleAttr** dataclass?

```python
            samp = Sample(
                date=s_attr.date,
                member=s_attr.membeer,
                input_terms=s_attr.input_terms,
                output_terms=s_attr.output_terms,
            )
```

Or you could even pass the Sample class to your **run_through_timestamps** and let it build the sample at the end, this way you avoid the intermediate "fog".

```python
samples = run_through_timestamps(self.period, self.settings, Sample)
``

For Titan you can use functools.partial to prepare a callable with the "constant attribute".

```python
PartialSample = functools.partial(Sample, settings=self.settings, params=self.params, stats=stats, grid=self.grid)
samples = run_through_timestamps(self.period, self.settings, Sample)
``

Just some ideas, let me know.

Reply:
We were also thinking of a `timestamp` object, that would also be useful to keep the difference between poesy (a reforecast with forecasts spanning more than 24h) and titan (mainly a reanalysis where valid times are enough).

Reply:
Initially that was planned for a later PR, but it could be time to enforce it now ?

Reply:
Well I would say don't leave the codebase in an intermediary state because you never now it might stay like that. For now at least I would instanciate the samples inside the functions an return a list of samples.

Reply:
well put, we are on this